### ISSUE TITLE
Goomph starting OSGI application fails since goomph 3.16

### ISSUE THREAD

**Issue by frankbenoit** [diffplug/goomph#92]
This was working before.
After upgrading Goomph to 3.16/17 I get this exception:

Gradle 4.10.3
* Exception is:                                                                                                                                                               java.lang.IllegalStateException: Resolving configuration 'runtimeElements' directly is not allowed                                                                                    at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.assertResolvingAllowed(DefaultConfiguration.java:1060)                                               at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$1500(DefaultConfiguration.java:123)                                                           at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getSelectedArtifacts(DefaultConfiguration.java:1036)                     at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getFiles(DefaultConfiguration.java:1025)                                 at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.getFiles(DefaultConfiguration.java:444)                                                              at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration_Decorated.getFiles(Unknown Source)                                                                   at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolve(DefaultConfiguration.java:434)                                                               at com.diffplug.gradle.eclipserunner.EquinoxLaunchSource.calculateState(EquinoxLaunchSource.java:81)                                                                          at com.diffplug.gradle.eclipserunner.EquinoxLaunchSource.calculateState(EquinoxLaunchSource.java:36)                                                                          at com.diffplug.spotless.LazyForwardingEquality.state(LazyForwardingEquality.java:56)                                                                                         at com.diffplug.spotless.LazyForwardingEquality.toBytes(LazyForwardingEquality.java:85)                                                                                       at com.diffplug.spotless.LazyForwardingEquality.hashCode(LazyForwardingEquality.java:102)
--

Gradle 5.2
* Exception is:                                                                                                                                                               org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':headlessAppSetup'.                                                                                           at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:38)                                                             at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:63)                                                             at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:49)                                                                    at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:46)                                                                    at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)                                  at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)                                  at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)                                                             at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)                                                               at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)                                                               at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)                                                                  at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)                                                             at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)                                                                   at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)                                                                                     at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)                                             at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)                                             at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)                                    at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)                                    at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)                                                                       at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)                                                                       at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)                                                                         at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)                                                                 at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)                                                                             at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)                                                                     at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)                                                                                      at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)                                                              Caused by: java.lang.IllegalStateException: Resolving configuration 'runtimeElements' directly is not allowed                                                                         at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.assertIsResolvable(DefaultConfiguration.java:1199)                                                   at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveToStateOrLater(DefaultConfiguration.java:530)                                                 at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$2200(DefaultConfiguration.java:136)                                                           at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getSelectedArtifacts(DefaultConfiguration.java:1159)                     at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.getFiles(DefaultConfiguration.java:1148)                                 at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.getFiles(DefaultConfiguration.java:468)                                                              at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration_Decorated.getFiles(Unknown Source)                                                                   at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolve(DefaultConfiguration.java:458)                                                               at com.diffplug.gradle.eclipserunner.EquinoxLaunchSource.calculateState(EquinoxLaunchSource.java:81)                                                                          at com.diffplug.gradle.eclipserunner.EquinoxLaunchSource.calculateState(EquinoxLaunchSource.java:36)                                                                          at com.diffplug.spotless.LazyForwardingEquality.state(LazyForwardingEquality.java:56)                                                                                         at com.diffplug.spotless.LazyForwardingEquality.toBytes(LazyForwardingEquality.java:85)                                                                                       at com.diffplug.spotless.LazyForwardingEquality.hashCode(LazyForwardingEquality.java:102)                                                                                     at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionStateTaskExecuter.snapshotTaskInputProperties(ResolveBeforeExecutionStateTaskExecuter.java:129)              at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionStateTaskExecuter.createExecutionState(ResolveBeforeExecutionStateTaskExecuter.java:90)                      at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionStateTaskExecuter.execute(ResolveBeforeExecutionStateTaskExecuter.java:72)                                   at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)                                                                     at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:109)                                                at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionOutputsTaskExecuter.execute(ResolveBeforeExecutionOutputsTaskExecuter.java:67)                               at org.gradle.api.internal.tasks.execution.ResolveAfterPreviousExecutionStateTaskExecuter.execute(ResolveAfterPreviousExecutionStateTaskExecuter.java:46)                     at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:93)                                                           at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:45)                                                     at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:94)                                                 at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)                                                       at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)                                                                     at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)                                                             ... 24 more
--

Can you please check?
regards
Frank

**Comment by nedtwigg** [comment]
Can you confirm that `3.18.0-SNAPSHOT` fixes your problem?  Available now in `https://oss.sonatype.org/content/repositories/snapshots`

Once it's confirmed good I'll release `3.17.2`.

**Comment by nedtwigg** [comment]
I bet it has something to do with a complicated POM in one of your deps, where whatever jar contains `com.sun.istack.Pool` is only included for "JRE8" or some other specific runtime environment.  Probably gradle 4.x does a fat search that just includes everything, but gradle 5.x is being more granular and only including what is appropriate based on the runtime, but the flags that determine that runtime aren't being set correctly somewhere deep inside gradle.

My recommendation would be to find the missing jar (maybe https://mvnrepository.com/artifact/com.sun.istack/istack-commons-runtime?) and add it explicitly in the equinoxLaunch.

My guess above is based on:

- https://github.com/Enterprise-Content-Management/documentum-rest-client-java/issues/4
- https://github.com/ben-manes/gradle-versions-plugin/issues/127#issuecomment-277207795
- https://github.com/ben-manes/gradle-versions-plugin/issues/274#issuecomment-440467534

Let me know if adding the missing jar is a viable workaround.  If not we can keep digging before I publish `3.17.2`.

**Comment by frankbenoit** [comment]
Hi Ned,

I tried it by doing this:

```
equinoxLaunch {
	headlessAppSetup {
		source.addProject(project(':......'))

		// *** I added this line ***
		source.addMaven('com.sun.istack:istack-commons-runtime:3.0.8')

		launchTask 'myApp', {
			it.args = ['-consoleLog', '-application', 'my.application']
			it.workingDir = file(myWorkdir)
		}
	}
}
```

But this did not change the error behavior.

Then i added it into the buildscript:
```
buildscript {
	repositories {
		mavenCentral()
	}
	dependencies {
		classpath "com.sun.istack:istack-commons-runtime:3.0.8"
	}
}
```

This fixed it.
Now it works with Gradle 5.1.1 and 5.2.

Thanks Ned!

Frank

**Comment by nedtwigg** [comment]
Published as `3.17.2`.
